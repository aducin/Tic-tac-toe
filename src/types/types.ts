interface BoxBase {
    icon: string;
    iconWinning: string;
    value: string;
};

export interface Box extends BoxBase {
    assigned: boolean;
    winning: boolean;
};

export type BoxesProps = {
    boxes: Box[];
    finished: boolean;
    onBoxClicked: (id: number) => void;
    started: boolean;
};

export type BoxItemProps = {
    data: {
        id: number;
        item: Box,
    };
    onBoxClicked: (id: number) => void;
}

export type ButtonsProps = {
    onButtonClicked: (button: string) => void;
    started: boolean;
};

export type DetailsProps = {
    activePlayer: string;
    turnTimer: number;
    value: string;
}

export type GameComponentState = {
    gameTimer: number;
    message: string;
    success: boolean;
    turnTimer: number;
    warning: boolean;
};

export type GameHistory = {
    duration: number;
    finishedTime: string;
    moves: number;
    winner: string;
};

export interface GameType extends BoxBase {
    id: number;
};

export type HistoryAction = {
    duration: number;
    finishedTime: Date;
}

export type IsPossibleToWin = {
    first: number;
    second: number
};

export type MatchedRows = {
    first: number;
    second: number;
};

export type MessageProps = {
    message: string;
    onRemoveAlert?: () => void;
    success: boolean;
    warning: boolean;
};

type Player = {
    active: boolean;
    name: string;
}

export type PlayerAction = {
    property: string;
    value: string;
}

export type PlayersProps = {
    onInputChanged: (property: string, name: string) => void;
    player1: string;
    player2: string;
}

export type ReducerAction = {
    payload: {
        boxes?: Box[];
        player?: PlayerAction;
        scoresList?: HistoryAction;
    }
    type: string;
};

export interface ScoresProps {
    scores: GameHistory[] | [];
}

export interface StoreState extends ScoresProps {
    boxes: Box[];
    finished: boolean;
    moves: number;
    next: GameType;
    player1: Player;
    player2: Player;
    started: boolean;
};
