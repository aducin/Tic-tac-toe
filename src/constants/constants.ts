import { GameType } from '../interfaces/interfaces';

export const DEFAULT_PLAYER1_NAME = 'Player 1';
export const DEFAULT_PLAYER2_NAME = 'Player 2';
export const EMPTY_HISTORY = 'Game history is empty at the moment';
export const FOOTER_TEXT = 'Tic-tac-toe footer';
export const GAME_NOT_STARTED = 'Click the button to start the game';
export const GAME_OVER = 'Game is over - no winner';
export const GAME_SUCCEDED = (winner: string, duration: number) =>
    `${winner} has won in ${duration.toFixed(2)} seconds!`;
export const HEADER_TEXT = 'Tic-tac-toe header';
export const MESSAGE_HEADER_SUCCESS = 'CONGRATULATIONS!';
export const MESSAGE_HEADER_WARNING = 'WARNING!';
export const MESSAGE_TIME = 3000;
export const MOBILE_MAX_WIDTH = '768px';
export const PLAYER_FIRST_TYPE = 'O';
export const PLAYER_SECOND_TYPE = 'X';
export const ROWS_TO_WIN = 3;
export const SCORES_LIST_HEADER = 'Score results';
export const TIME_IS_OVER = 'Time is over - next move';
export const TURN_TIME = 30;
export const WINNING_ROWS = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
];
export const GAMETYPES: GameType[] = [
    { 
        id: 1,
        icon: 'ImageO',
        iconWinning: 'ImageOWinning',
        value: PLAYER_FIRST_TYPE
    },
    {
        id: 2,
        icon: 'ImageX',
        iconWinning: 'ImageXWinning',
        value: PLAYER_SECOND_TYPE
    }
];
