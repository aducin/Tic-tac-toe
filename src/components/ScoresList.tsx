import React from 'react';
import ScoresRow from '../components/ScoresRow';
import { GameHistory, ScoresProps } from '../interfaces/interfaces';
import { MOBILE_MAX_WIDTH } from '../constants/constants';
import { useMediaQuery } from 'react-responsive';
import styles from '../assets/styles/scoresList.module.scss';

const ScoresList: React.FC<ScoresProps> = (props: ScoresProps) => {
    const isMobile = useMediaQuery({ query: `(max-width: ${MOBILE_MAX_WIDTH})` });
    const list = (props.scores as (GameHistory)[]).map((el: GameHistory, index: number) => {
        return isMobile ?
            (
                <li key={index} className={styles.mobileItem}>
                    <ScoresRow label="Winner" styles={styles.mobileItemRow} value={el.winner} />
                    <ScoresRow label="Moves" styles={styles.mobileItemRow} value={el.moves} />
                    <ScoresRow label="Game time" styles={styles.mobileItemRow} value={`${el.duration.toFixed(2)}sec.`} />
                    <ScoresRow label="Date" styles={styles.mobileItemRow} value={el.finishedTime} />
                </li>
            ) :
            (
                <tr key={index}>
                    <th>{el.winner}</th>
                    <th>{el.moves}</th>
                    <th>{el.duration.toFixed(2)}sec.</th>
                    <th>{el.finishedTime}</th>
                </tr>
            );
    }); 

    const content = isMobile ?
        (
            <ul className={styles.mobileList}>
                {list}
            </ul>
        ) :
        (
            <table className="table table-striped">
                <thead>
                    <tr>
                        <th>Winner</th>
                        <th>Moves</th>
                        <th>Game time</th>
                        <th>Date</th>
                    </tr>
                </thead>
                <tbody>
                    {list}
                </tbody>
            </table>
        );
    
    return(
        <div className={styles.container}>
            {content}
        </div>
    );
};

export default React.memo(ScoresList);
