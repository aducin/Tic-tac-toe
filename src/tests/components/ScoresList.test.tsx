import React from "react";
import TestRenderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { ShallowWrapper } from 'enzyme';
import ScoresList from '../../components/ScoresList';
import { ScoresProps } from '../../types/types';
import { DEFAULT_PLAYER1_NAME, DEFAULT_PLAYER2_NAME } from '../../constants/constants';

const data: ScoresProps = {
    scores: [
        {
            duration: 1.577,
            finishedTime: "2020-03-19 10:00",
            moves: 5,
            winner: DEFAULT_PLAYER1_NAME
        },
        {
            duration: 2.252,
            finishedTime: "2020-03-20 10:15",
            moves: 5,
            winner: DEFAULT_PLAYER2_NAME
        }
    ]
}

const component = <ScoresList scores={data.scores} />;

const testData = TestRenderer.create(component);

describe('ScoresList component snapshot', () => {
    it('ScoresList component renders correctly', async () => {
        expect(testData.toJSON()).toMatchSnapshot();
    });
});

describe('AccountDetails tests', () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        wrapper = shallow(component);
    });

    it('there is a table with two rows', () => {
        expect(wrapper.find('table')).toBeTruthy;
        expect(wrapper.find('tbody').first().children()).toHaveLength(2);
    });

    it('there should be a correct duration data displayed in each row', () => {
        const el = wrapper.find('tbody').first().children();
        expect(el.first().children().first().text()).toEqual(DEFAULT_PLAYER1_NAME);
        expect(el.last().children().first().text()).toEqual(DEFAULT_PLAYER2_NAME);
    });
});
