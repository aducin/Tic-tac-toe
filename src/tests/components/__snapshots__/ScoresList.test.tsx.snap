// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoresList component snapshot ScoresList component renders correctly 1`] = `
<div
  className="container"
>
  <table
    className="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Winner
        </th>
        <th>
          Moves
        </th>
        <th>
          Game time
        </th>
        <th>
          Date
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          Player 1
        </th>
        <th>
          5
        </th>
        <th>
          1.58
          sec.
        </th>
        <th>
          2020-03-19 10:00
        </th>
      </tr>
      <tr>
        <th>
          Player 2
        </th>
        <th>
          5
        </th>
        <th>
          2.25
          sec.
        </th>
        <th>
          2020-03-20 10:15
        </th>
      </tr>
    </tbody>
  </table>
</div>
`;
